The following builds have failed:
- [Quick Feedback - Linux Only (Trigger) (Check)](https://builds.gradle.org/buildConfiguration/Gradle_Master_Check_Stage_QuickFeedbackLinuxOnly_Trigger/115683121)
  - [Build Scan](https://ge.gradle.org/scans?search.buildOutcome=failure&amp;search.tags=CI&amp;search.names=gitCommitId&amp;search.values=581edd389ba82f5bf732ced69bf00c28c471ad6c&amp;)

<details><summary>🔎 3 failure groups (0 build, 3 test)</summary>

🧪 **`ConfigurationCacheBuildCacheIntegrationTest.in warning mode, cacheable task is cached and configuration reused despite execution time problem`** in `:build-cache:embeddedIntegTest` · 255 occurrences · [scan](https://ge.gradle.org/s/uiduiywhpi3tu/tests/overview?outcome=failed)
> Condition failed with Exception:
> 
> problems.assertResultHasProblems(result) { withInput("$location: file '$testFileName'") ignoringUnexpectedInputs() }
> |        |                       |
> |        |                       <org.gradle.integtests.fixtures.executer.InProcessGradleExecuter$InProcessExecutionResult@3c34abb delegate=org.gradle.integtests.fixtures.executer.OutputScrapingExecutionResult@2334b42a executedTasks=[:help] skippedTasks=[]>
> |        java.lang.AssertionError: 
> |        Expected: n…
🧪 **Test failure** · 1 occurrence · [scan](https://ge.gradle.org/s/6sb3o7zqcjaca/tests/overview?outcome=failed)
> Multiple Failures (7 failures)
> 	org.spockframework.runtime.SpockComparisonFailure: Condition not satisfied:
> 
> fqid == REGISTRATION_UNSUPPORTED
> |    |  |
> |    |  validation:configuration-cache:registration-of-listener-on-gradle-buildfinished-is-unsupported
> |    false
> |    49 differences (47% similarity)
> |    validation:configuration-cache:(conf)i(gu)ration-(cach---)e(---)-(wa--)r(------------)n(-----)-(m--------)o(d-)e(-)…
🧪 **Test failure** · 1 occurrence · [scan](https://ge.gradle.org/s/6sb3o7zqcjaca/tests/overview?outcome=failed)
> Multiple Failures (2 failures)
> 	org.spockframework.runtime.SpockComparisonFailure: Condition not satisfied:
> 
> fqid == REGISTRATION_UNSUPPORTED
> |    |  |
> |    |  validation:configuration-cache:registration-of-listener-on-gradle-buildfinished-is-unsupported
> |    false
> |    49 differences (47% similarity)
> |    validation:configuration-cache:(conf)i(gu)ration-(cach---)e(---)-(wa--)r(------------)n(-----)-(m--------)o(d-)e(-)…
</details>

<sub>Was this summary helpful? [Good 👍](https://gradle-bot.grdev.net/feedback?pr=38320&build=115683121&rating=positive) · [Bad 👎](https://gradle-bot.grdev.net/feedback?pr=38320&build=115683121&rating=negative) · or [report an issue in #bt-dev-prod](https://gradle.slack.com/archives/CBSJ2GUTV)</sub>

--- a/platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/AbstractConfigurationCacheIntegrationTest.groovy
+++ b/platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/AbstractConfigurationCacheIntegrationTest.groovy
@@ -20,6 +20,7 @@ import org.gradle.initialization.StartParameterBuildOptions.ConfigurationCacheMa
 import org.gradle.initialization.StartParameterBuildOptions.ConfigurationCacheOption
 import org.gradle.initialization.StartParameterBuildOptions.ConfigurationCacheParallelOption
 import org.gradle.initialization.StartParameterBuildOptions.ConfigurationCacheReadOnlyOption
+import org.gradle.integtests.fixtures.problems.ReceivedProblem
 import org.gradle.internal.cc.impl.fixtures.AbstractConfigurationCacheOptInFeatureIntegrationTest
 
 abstract class AbstractConfigurationCacheIntegrationTest extends AbstractConfigurationCacheOptInFeatureIntegrationTest {
@@ -48,6 +49,13 @@ abstract class AbstractConfigurationCacheIntegrationTest extends AbstractConfigu
 
     void configurationCacheRunLenient(String... tasks) {
         run(WARN_PROBLEMS_CLI_OPT, *CLI_OPTIONS, *tasks)
+        if (isProblemsApiCheckEnabled()) {
+            consumeWarnModeProblem()
+        }
+    }
+
+    protected ReceivedProblem consumeWarnModeProblem() {
+        return findReceivedProblem { it.fqid.endsWith(":configuration-cache-warn-mode") }
     }
 
     void configurationCacheFails(String... tasks) {

--- a/platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/ConfigurationCacheEnablementIntegrationTest.groovy
+++ b/platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/ConfigurationCacheEnablementIntegrationTest.groovy
@@ -66,24 +66,13 @@ class ConfigurationCacheEnablementIntegrationTest extends AbstractConfigurationC
 
     def "configuration cache warn mode is reported through the Problems API"() {
         given:
-        buildFile "task ok {}"
         enableProblemsApiCheck()
 
         when:
-        run("ok", "--configuration-cache", "--configuration-cache-problems=warn")
+        run("help", "--configuration-cache", "--configuration-cache-problems=warn")
 
         then:
-        receivedProblems.size() == 1
-
-        verifyAll(receivedProblem) {
-            fqid == 'validation:configuration-cache:configuration-cache-warn-mode'
-
-            contextualLabel == 'Configuration Cache warn mode is ENABLED. ' +
-                'Configuration Cache problems are being ignored. ' +
-                'Build outputs may be incorrect and the build may crash unexpectedly. ' +
-                'Use this only to discover configuration cache problems. ' +
-                'Do not use this to produce artifacts.'
-
+        verifyAll(consumeWarnModeProblem()) {
             definition.severity == Severity.WARNING
         }
     }

--- a/platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/ConfigurationCacheProblemsServiceIntegTest.groovy
+++ b/platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/ConfigurationCacheProblemsServiceIntegTest.groovy
@@ -49,8 +49,7 @@ class ConfigurationCacheProblemsServiceIntegTest extends AbstractConfigurationCa
         }
 
         and:
-        verifyAll(receivedProblem(0)) {
-            fqid == REGISTRATION_UNSUPPORTED
+        verifyAll(findReceivedProblem { it.fqid == REGISTRATION_UNSUPPORTED }) {
             contextualLabel == "registration of listener on 'Gradle.buildFinished' is unsupported"
             definition.severity == Severity.WARNING
             definition.documentationLink.url.endsWith("/userguide/configuration_cache_requirements.html#config_cache:requirements:build_listeners")
@@ -65,7 +64,7 @@ class ConfigurationCacheProblemsServiceIntegTest extends AbstractConfigurationCa
         configurationCacheRunLenient 'run', "-D${StartParameterBuildOptions.ConfigurationCacheRecreateOption.PROPERTY_NAME}=true"
 
         then:
-        verifyAll(receivedProblem) {
+        verifyAll(findReceivedProblem { it.fqid == REGISTRATION_UNSUPPORTED }) {
             fqid == REGISTRATION_UNSUPPORTED
             contextualLabel == "registration of listener on 'Gradle.buildFinished' is unsupported"
             definition.severity == Severity.WARNING
@@ -88,10 +87,10 @@ class ConfigurationCacheProblemsServiceIntegTest extends AbstractConfigurationCa
 
         when:
         configurationCacheFails WARN_PROBLEMS_CLI_OPT, "-D$MAX_PROBLEMS_GRADLE_PROP=0", 'run'
+        consumeWarnModeProblem()
 
         then:
-        verifyAll(receivedProblem) {
-            fqid == REGISTRATION_UNSUPPORTED
+        verifyAll(findReceivedProblem { it.fqid == REGISTRATION_UNSUPPORTED }) {
             contextualLabel == "registration of listener on 'Gradle.buildFinished' is unsupported"
             definition.severity == Severity.WARNING
         }

--- a/platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/ConfigurationCacheProblems.kt
+++ b/platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/ConfigurationCacheProblems.kt
@@ -83,9 +83,9 @@ val isolatedProjectsDangerouslyIgnoreProblemsSentences = listOf(
 
 val configurationCacheWarnModeSentences = listOf(
     "Configuration Cache warn mode is ENABLED.",
-    "Configuration Cache problems are being ignored.",
+    "Configuration Cache constraint violations are being ignored.",
     "Build outputs may be incorrect and the build may crash unexpectedly.",
-    "Use this only to discover configuration cache problems.",
+    "Use this only to discover configuration cache incompatibilities.",
     "Do not use this to produce artifacts."
 )
 

--- a/testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/AbstractIntegrationSpec.groovy
+++ b/testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/AbstractIntegrationSpec.groovy
@@ -859,6 +859,19 @@ tmpdir is currently ${System.getProperty("java.io.tmpdir")}""")
     }
 
     /**
+     * Returns the first received problem that matches the given criteria and marks it as validated.
+     * The validation should only verify the contextual attributes of the received problem, as the problem definition is checked at the end of each test (see {@see KnownProblemIds}).
+     * Fails if no problem matches the condition.
+     *
+     * @see #receivedProblem(int)
+     */
+    protected ReceivedProblem findReceivedProblem(Closure<ReceivedProblem> criteria) {
+        def found = getReceivedProblems().findIndexOf { it?.collect(criteria) }
+        assert found >= 0 : "No problems match the criteria"
+        return receivedProblem(found)
+    }
+
+    /**
      * Generates a `repositories` block pointing to the standard maven repo fixture.
      *
      * This is often required for running with configuration cache enabled, as

--- a/testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/AbstractIntegrationSpec.groovy
+++ b/testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/AbstractIntegrationSpec.groovy
@@ -55,6 +55,7 @@ import org.opentest4j.AssertionFailedError
 import spock.lang.Specification
 
 import java.nio.file.Files
+import java.util.function.Predicate
 import java.util.regex.Pattern
 
 import static org.gradle.integtests.fixtures.timeout.IntegrationTestTimeout.DEFAULT_TIMEOUT_SECONDS
@@ -865,10 +866,10 @@ tmpdir is currently ${System.getProperty("java.io.tmpdir")}""")
      *
      * @see #receivedProblem(int)
      */
-    protected ReceivedProblem findReceivedProblem(Closure<ReceivedProblem> criteria) {
-        def found = getReceivedProblems().findIndexOf { it?.collect(criteria) }
-        assert found >= 0 : "No problems match the criteria"
-        return receivedProblem(found)
+    protected ReceivedProblem findReceivedProblem(Predicate<ReceivedProblem> criteria) {
+        def index = getReceivedProblems().findIndexOf { it && criteria.test(it) }
+        assert index >= 0: "No received problem matches the given criteria"
+        return receivedProblem(index)
     }
 
     /**

